cmd/snap/cmd_routine_portal_info.go

```diff
@@ -0,0 +1,148 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+	"text/template"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/sandbox/cgroup"
+)
+
+type cmdRoutinePortalInfo struct {
+	clientMixin
+	PortalInfoOptions struct {
+		Pid int
+	} `positional-args:"true" required:"true"`
+}
+
+var shortRoutinePortalInfoHelp = i18n.G("Return information about a process")
+var longRoutinePortalInfoHelp = i18n.G(`
+The portal-info command returns information about a process in keyfile format.
+
+This command is used by the xdg-desktop-portal service to retrieve
+information about snap confined processes.
+`)
+
+func init() {
+	addRoutineCommand("portal-info", shortRoutinePortalInfoHelp, longRoutinePortalInfoHelp, func() flags.Commander {
+		return &cmdRoutinePortalInfo{}
+	}, nil, []argDesc{{
+		// TRANSLATORS: This needs to begin with < and end with >
+		name: i18n.G("<process ID>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Process ID of confined app"),
+	}})
+}
+
+const portalInfoTemplate = `[Snap Info]
+InstanceName={{.Snap.Name}}
+{{- if .App}}
+AppName={{.App.Name}}
+{{- end}}
+{{- if .DesktopFile}}
+DesktopFile={{.DesktopFile}}
+{{- end}}
+HasNetwork={{.HasNetwork}}
+`
+
+var (
+	cgroupSnapNameFromPid   = cgroup.SnapNameFromPid
+	apparmorSnapNameFromPid = apparmor.SnapNameFromPid
+)
+
+func (x *cmdRoutinePortalInfo) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	snapName, err := cgroupSnapNameFromPid(x.PortalInfoOptions.Pid)
+	if err != nil {
+		return err
+	}
+	snap, _, err := x.client.Snap(snapName)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve info for snap %q: %v", snapName, err)
+	}
+
+	// Try to identify the application name from AppArmor
+	var app *client.AppInfo
+	if snapName, appName, _, err := apparmorSnapNameFromPid(x.PortalInfoOptions.Pid); err == nil && snapName == snap.Name && appName != "" {
+		for i := range snap.Apps {
+			if snap.Apps[i].Name == appName {
+				app = &snap.Apps[i]
+				break
+			}
+		}
+	}
+	// As a fallback, pick an app with a desktop file, favouring
+	// the app named identically to the snap.
+	if app == nil {
+		for i := range snap.Apps {
+			if snap.Apps[i].DesktopFile != "" && (app == nil || snap.Apps[i].Name == snap.Name) {
+				app = &snap.Apps[i]
+			}
+		}
+	}
+
+	var desktopFile string
+	if app != nil {
+		desktopFile = filepath.Base(app.DesktopFile)
+	}
+
+	// Determine whether the snap has access to the network
+	connections, err := x.client.Connections(&client.ConnectionOptions{
```

Comment:
maybe have a TODO about having a direct API for this (we do is-connected nowadays on the snapctl side but not something equivalent general API side)

Reply:
added